snooty/postprocess.py

```diff
@@ -1552,36 +1552,48 @@ def _attach_doc_title(self, fileid_stack: FileIdStack, node: n.RefRole) -> None:
 class FacetsHandler(Handler):
     """Builds page.facets depending on facets found on nodes on this page"""
 
-    # TODO: should be able to handle facet propagation from project/directory/page -> page in the future
-    # when taxonomy and function of facets (propagation) are decided
     def __init__(self, context: Context) -> None:
         super().__init__(context)
 
-        self.facets: SerializedNode = {}
-        self.target: SerializedNode = self.facets
+        self.facets: List[Facet] = []
+        self.parent_stack: List[Tuple[Facet, int]] = []
         self.removal_nodes: List[n.Node] = []
 
     def enter_node(self, fileid_stack: FileIdStack, node: n.Node) -> None:
         if not isinstance(node, n.Directive) or node.name != "facet":
             return
 
-        facet_node: SerializedNode = {"name": node.options.get("values", "")}
-        if not self.target.get(node.options["name"]):
-            self.target[node.options["name"]] = []
-        target_list = self.target[node.options["name"]]
-        if isinstance(target_list, list):
-            target_list.append(facet_node)
+        facet_node = Facet(category=node.options["name"], value=node.options["values"])
+
+        if self.parent_stack:
+            parent = self.parent_stack[-1][0]
+            if isinstance(parent.sub_facets, list):
```

Comment:
sub_facets is an `Optional[list]` right? Much better if that's the case to just have
```
sub_facets = parent.sub_facets
if sub_facets is not None:
    sub_facets.append(facet_node)
```